date-time/src/commonMain/kotlin/com/splendo/kaluga/datetime/timer/MonotonicTimer.kt

```diff
@@ -0,0 +1,209 @@
+                elapsedSoFar: Duration,
+                override val totalDuration: Duration
+            ) : NotRunning(elapsedSoFar), Timer.State.NotRunning.Paused {
+                fun start(
+                    interval: Duration,
+                    tickCorrection: TickCorrection,
+                    coroutineScope: CoroutineScope,
+                    finishCallback: suspend () -> Unit
+                ): Running = Running(
+                    elapsedSoFar = elapsed,
+                    totalDuration = totalDuration,
+                    interval = interval,
+                    tickCorrection = tickCorrection,
+                    coroutineScope = coroutineScope,
+                    finishCallback = finishCallback
+                )
+                internal fun finish(): Finished = Finished(totalDuration)
+            }
+
+            /** Timer is finished. */
+            class Finished(
+                override val totalDuration: Duration
+            ) : NotRunning(totalDuration), Timer.State.NotRunning.Finished
+        }
+
+        /** Timer is running. */
+        class Running(
+            elapsedSoFar: Duration,
+            override val totalDuration: Duration,
+            interval: Duration,
+            tickCorrection: TickCorrection,
+            private val coroutineScope: CoroutineScope,
+            private val finishCallback: suspend () -> Unit
+        ) : State(), Timer.State.Running, HandleAfterNewStateIsSet<State>, HandleBeforeOldStateIsRemoved<State> {
+            override val elapsed: Flow<Duration> = tickProvider(
+                offset = elapsedSoFar,
+                max = totalDuration,
+                interval = interval,
+                correction = tickCorrection
+            )
+            private lateinit var finishJob: Job
+
+            internal suspend fun pause(): NotRunning.Paused {
+                return NotRunning.Paused(elapsed.first(), totalDuration)
+            }
+            internal fun finish(): NotRunning.Finished = NotRunning.Finished(totalDuration)
+
+            override suspend fun beforeOldStateIsRemoved(oldState: State) {
+                finishJob = coroutineScope.launch {
+                    delay(totalDuration - elapsed.first())
+                    finishCallback()
+                }
+            }
+            override suspend fun afterNewStateIsSet(newState: State) {
+                finishJob.cancel()
+            }
+        }
+    }
+}
+
+private fun tickProviderWithOffsetCorrection(
+    offset: Duration,
+    max: Duration,
+    interval: Duration,
+    correctionOffset: Duration
+): Flow<Duration> {
+    val mark = TimeSource.Monotonic.markNow()
+    return flow {
+        while (true) {
+            val elapsed = offset + mark.elapsedNow()
+            emit(elapsed.coerceAtMost(max))
+            if (max <= elapsed) break
+
+            delay(interval + correctionOffset)
+        }
+    }
+}
+
+internal fun tickProviderWithAdaptiveCorrection(
```

Comment:
there is a slight problem that timer might undershoot from time to time. e.g: user wanted a timer which updates every second - a possible output is
```
2ms
999ms
2002ms
3005ms
3999ms
```
it'd create a problem if user truncates the time and not rounds it.